cardano-testnet/src/Testnet/Components/SPO.hs

```diff
@@ -0,0 +1,327 @@
+  , registerSingleSpo
+  ) where
+
+import           Cardano.Api.Shelley hiding (cardanoEra)
+
+import           Control.Monad
+import           Control.Monad.Catch (MonadCatch)
+import           Control.Monad.IO.Class (MonadIO)
+import qualified Data.Aeson as Aeson
+import qualified Data.Map.Strict as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import           GHC.Stack (HasCallStack)
+import qualified GHC.Stack as GHC
+import           System.FilePath.Posix ((</>))
+
+import           Hedgehog
+import           Hedgehog.Extras (ExecConfig, threadDelay)
+import qualified Hedgehog.Extras.Test.Base as H
+import qualified Hedgehog.Extras.Test.File as H
+
+import           Testnet.Process.Cli
+import           Testnet.Process.Run (execCli, execCli', execCli_)
+import           Testnet.Runtime
+import           Testnet.Start.Types
+
+checkStakePoolRegistered
+  :: (MonadTest m, MonadCatch m, MonadIO m, HasCallStack)
+  => TmpAbsolutePath
+  -> ExecConfig
+  -> FilePath -- ^ Satke pool cold verification key file
+  -> CardanoTestnetOptions
+  -> FilePath -- ^ Output file path of stake pool info
+  -> m String
+checkStakePoolRegistered tempAbsPath execConfig poolColdVkeyFp cTestnetOpts outputFp =
+  GHC.withFrozenCallStack $ do
+    let tempAbsPath' = unTmpAbsPath tempAbsPath
+        testnetMag = cardanoTestnetMagic cTestnetOpts
+        oFpAbs = tempAbsPath' </> outputFp
+
+    stakePoolId' <- filter ( /= '\n') <$>
+                      execCli [ "stake-pool", "id"
+                              , "--cold-verification-key-file", poolColdVkeyFp
+                              ]
+
+    -- Check to see if stake pool was registered
+    void $ execCli' execConfig
+      [ "query", "stake-pools"
+      , "--testnet-magic", show @Int testnetMag
+      , "--out-file", oFpAbs
+      ]
+
+    currRegPools <- H.leftFailM $ H.readJsonFile oFpAbs
+    poolIds <- H.noteShowM $ H.jsonErrorFail $ Aeson.fromJSON @(Set PoolId) currRegPools
+    H.note_ "Check stake pool was successfully registered"
+
+    let stakePoolNotRegisteredFailure = unlines [ "Stake pool was not registered"
+                                                , "Expected to be registed: ", show stakePoolId'
+                                                , "Currently registered: ", show (Set.toList $ Set.map serialiseToBech32 poolIds)
+                                                ]
+    if Set.member stakePoolId' $ Set.map (Text.unpack . serialiseToBech32) poolIds
+    then return stakePoolId'
+    else H.failWithCustom GHC.callStack Nothing stakePoolNotRegisteredFailure
+
+checkStakeKeyRegistered
+  :: (MonadTest m, MonadCatch m, MonadIO m, HasCallStack)
+  => TmpAbsolutePath
+  -> ExecConfig
+  -> CardanoTestnetOptions
+  -> String -- ^ Stake address
+  -> FilePath -- ^ Output file path of stake address info
+  -> m DelegationsAndRewards
+checkStakeKeyRegistered tempAbsPath execConfig cTestnetOpts stakeAddr outputFp  =
```

Comment:
Why is this called `checkStakeKeyRegistered` and not something like `getStakeDelegrationAndRewards`?

Reply:
Because we fail if the stake key isn't found:

```
dag@(DelegationsAndRewards (rewardsMap, _delegMap)) <- H.noteShowM $ H.jsonErrorFail $ Aeson.fromJSON @DelegationsAndRewards pledgerStakeInfo
    case Map.lookup sAddr rewardsMap of
      Nothing -> H.failWithCustom GHC.callStack Nothing
                   $ unlines [ "Stake address: "
                             , Text.unpack (serialiseToBech32 sAddr)
                             , "was not registered"
                             , "Current registered stake keys: "
                             , show $ map serialiseToBech32 $ Map.keys rewardsMap
                             ]
      Just _ -> return dag
```

Reply:
`checkStakeKeyRegisteredAndReturnRewards`?